### ISSUE TITLE
Error on SELECT FOR XML statement

### ISSUE THREAD

**Issue by SillyDillyWilly** [erikdarlingdata/DarlingData#191]
version = '2.50', 
version_date = '20220701';

SQL Server 2014

I'm currently quite often getting this error message while the last query in @cpu_sql is being executed:

Msg 6841, Level 16, State 1, Line 6
FOR XML could not serialize the data for node 'processing-instruction(query)' because it contains a character (0x0000) which is not allowed in XML. To retrieve this data using FOR XML, convert it to binary, varbinary or image data type and use the BINARY BASE64 directive.

**Comment by erikdarlingdata** [comment]
Thanks. I'll work on this soon.

**Comment by erikdarlingdata** [comment]
@SillyDillyWilly can you try the version of sp_PressureDetector in the dev branch?

**Comment by SillyDillyWilly** [comment]
I get a compilation error when I try to install the new code on my SQL Server 2014 instance:

Msg 207, Level 16, State 1, Procedure sp_PressureDetector, Line 788 [Batch Start Line 46]
Invalid column name 'target_kb'.

I don't get the compilation error on a SQL Server 2016, so that must be relateted to a version-dependent change.

**Comment by SillyDillyWilly** [comment]
I tried to change the code section that couldn't compile on SQL Server 2014 into a dynamic query surrounded by a rudimentory try catch construct and that allowed the sp to be compiled.

`        IF @azure = 1
        BEGIN
            /*
            SELECT 
                @total_physical_memory_gb =
                    SUM(domn.target_kb / 1024.) 
            FROM sys.dm_os_memory_nodes AS domn;
            */
            BEGIN TRY
                EXEC sp_executesql 
                    @stmt = N'
                            SELECT 
                                @total_physical_memory_gb =
                                    SUM(domn.target_kb / 1024.) 
                            FROM sys.dm_os_memory_nodes AS domn;',
                    @params = N'@total_physical_memory_gb bigint OUTPUT',
                    @total_physical_memory_gb = @total_physical_memory_gb OUTPUT
            END TRY
            BEGIN CATCH
            END CATCH
        END;`

However when I run the sp the originally reported error still shows up:
Msg 6841, Level 16, State 1, Line 6
FOR XML could not serialize the data for node 'processing-instruction(query)' because it contains a character (0x0000) which is not allowed in XML. To retrieve this data using FOR XML, convert it to binary, varbinary or image data type and use the BINARY BASE64 directive.

So no luck with the code in dev either, unfortunately.

**Comment by HannahVernon** [comment]
@erikdarlingdata - from testing 2014, this seems like a possible alternate to obtain the target memory quantity:

```
SELECT 
    @total_physical_memory_gb = 
        SUM(dosi.[committed_target_kb]) / 1024.
FROM [sys].[dm_os_sys_info] dosi
```

On my SQL Server 2019 box, both versions return the same quantity.  I'm also working on the other issue identified above to see if I can repro it.

**Comment by HannahVernon** [comment]
I've been unable to repro the XML serialization issue above.  Tried on SQL Server 2014, SQL Server 2017, and SQL Server 2019.  I do see the code is replacing `NCHAR(0)` with `?`, so theoretically this shouldn't be an issue.

**Comment by SillyDillyWilly** [comment]
The [REPLACE documentation](https://learn.microsoft.com/en-us/sql/t-sql/functions/replace-transact-sql?view=sql-server-ver16) says:

_REPLACE performs comparisons based on the collation of the input. To perform a comparison in a specified collation, you can use [COLLATE](https://learn.microsoft.com/en-us/sql/t-sql/statements/collations?view=sql-server-ver16) to apply an explicit collation to the input.

0x0000 (char(0)) is an undefined character in Windows collations and cannot be included in REPLACE._

**Comment by HannahVernon** [comment]
Any chance you can upload a sample of the problematic xml, and share it with us?

**Comment by erikdarlingdata** [comment]
Yeah, if the big REPLACE still doesn’t fix it, we’d need sample query text to work with. If you can’t provide one, then you’re going to have to figure out the problem in your own and send a pull request over to fix it. I also can’t repro the issue locally.

I’m going to leave this issue open to work on and track the dynamic SQL issue.

/*
I typed this on a phone.
Thanks for understanding.
*/
________________________________

**Comment by SillyDillyWilly** [comment]
@HannahVernon: Sorry, I'm allowed to share the queries with the class.

But... I tried to change the collation on the last REPLACE in the sequence to NCHAR(1),N''?'') COLLATE Latin1_General_BIN,NCHAR(0),N''''), i.e. I applied a binary collation, which enables the replace function to handle the 0x0000 character.

After this change I have not seen the error, but I will of course continue to check, and report back if I see it surface again.

I'm not sure if replacing NCHAR(0) with a zero-length-string is correct. It seems to me that this might introduce a possible error with the SUBSTRING based on the provided start-offset. Maybe replacing it with ? like the others might be more secure?

**Comment by erikdarlingdata** [comment]
@SillyDillyWilly can you post the exact code fragment you used here? Or create a pull request with the change that worked for you so we can look at it? Thanks!

**Comment by SillyDillyWilly** [comment]
Sorry, I thought I had done that already, but I will try to be a little more succinct and precise in my comments. :-)
(Most of the previous comments relate the code in the dev branch, which might have contributed to your confusion).

In order to combat the reported error, the bare minimum of what is needed to be changed in the production code, i.e. the code in the main branch, is:

Line 907: dest.text,
change it to REPLACE(dest.text  COLLATE Latin1_General_BIN, NCHAR(0), N''?'')

The reason for the binary collation on the REPLACE is that in the normal windows text collations, char(0)/nchar(0) is an undefined character. This means that a simple REPLACE will not work. You need to use a binary collation for this to work for a 0x0000 character.

However, the above change apparently resulted in a 4000 character boundary being broken along the way, because some code (in the middle of the concatenated string!) was suddenly cut off. The result was an error when the procedure was called.

I think that in order to prevent this completely, you need to cast all concatenated elements to varchar(max) - at least if you use the + sign to do the concatenation.

To prevent this phenomenon, I changed the code to use the CONCAT function instead. If any of the elements are of type varchar(max) all elements are automatically converted to varchar(max) before the elements are concatenated. This makes for a much better control of the process than using the + for concatenation, IMO.


I have made my first-ever pull request and hope that it went well. :-)

**Comment by erikdarlingdata** [comment]
Thanks. I’ll take a look! Don’t take offense if I cherry pick some of your changes out and add them into the current dev branch.

/*
I typed this on a phone.
Thanks for understanding.
*/
________________________________

**Comment by erikdarlingdata** [comment]
Okay, got a version up in dev that should take care of it. I have to skip on the concat change, since I still run into folks using pre-2012 boxes that I need to support.

**Comment by erikdarlingdata** [comment]
Yeah, high-minded times, those were.

**Comment by SillyDillyWilly** [comment]
So maybe you should amend your ReadMe file? :-)

I have an idea that could be an improvement: 
Currently some of the listed queries from your sp looks like this:
FETCH API_CURSOR000000000075BD67

That isn't really all that helpful as far as showing what is actually going on. Would it be possible to add the actual running query, i.e. FETCH API_CURSOR000000000075BD67 : SELECT ... ?

The only way that I know of how to see the actual query that is running is to use the session id like this:
DECLARE @SessionId int = 756
SELECT c.session_id, c.properties, c.creation_time, c.is_open, t.text 
FROM sys.dm_exec_cursors (@SessionId) c 
CROSS APPLY sys.dm_exec_sql_text (c.sql_handle) t

This shows information like this:
session_id  properties                                   creation_time           is_open text
----------- -------------------------------------------- ----------------------- ------- --------------------------------------------------------------
756         API | Fast_Forward | Read Only | Global (0)  2022-11-17 10:46:32.070 1       (@P1 varchar(1000),@P2 varchar(1000),@P3 varchar(1000))SELECT ...

What do you think?

**Comment by erikdarlingdata** [comment]
I think that's a separate issue, and you should open it as one.
